### PR TITLE
feat(extender): Expose and reuse login and logout methods

### PR DESCRIPTION
**Pull request by zFernand0**
## Proposed changes

Expose the login and logout methods to extenders

## Release Notes

Milestone: 2.14.0

Changelog: Exposed the login and logout methods to extenders

## Types of changes

What types of changes does your code introduce to Zowe Explorer?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Updates to Documentation or Tests (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This checklist will be used as reference for both the contributor and the reviewer_

- [x] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/vscode-extension-for-zowe/wiki/Best-Practices:-Contributor-Guidance) wiki
- [x] PR title follows [Conventional Commits Guidelines](https://www.conventionalcommits.org/en/v1.0.0-beta.2/)
- [x] PR Description is included
- [ ] gif or screenshot is included if visual changes are made
- [ ] `yarn workspace vscode-extension-for-zowe vscode:prepublish` has been executed
- [ ] All checks have passed (DCO, Jenkins and Code Coverage)
- [ ] I have added unit test and it is passing
- [ ] I have added integration test and it is passing
- [ ] There is coverage for the code that I have added
- [ ] I have tested it manually and there are no regressions found
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any PR dependencies have been merged and published (if appropriate)

## Further comments

There is a dependency on the PR below:
- https://github.com/zowe/zowe-cli/pull/1983

### CODE REVIEW

**Review by JillieBeanSim — DISMISSED** [review]
Hey @zFernand0, thanks for this new extender functionality.  As regression testing with an extender that logs into it own rest api worked fine but when I tried with zosmf / apiml combo I got this error that I don't see with released ZE using same profile
Screen Shot 2023-12-15 at 9 45 51 AM

**Comment by JillieBeanSim** [comment]
further info for recreation steps of getting the error:

have team config with zosmf & base profile similar to below:
        "apimlzosmf": {
            "type": "zosmf",
            "properties": {
                "port": 7554,
                "basePath": "ibmzosmf/api/v1"
            },
            "secure": []
        },
       "base": {
            "type": "base",
            "properties": {
                "host": "host.com",
                "rejectUnauthorized": false
            },
            "secure": []
        }

On this branch I get the error above logging in, but works fine using the released ZE from marketplace.

**Comment by zFernand0** [comment]
> Hey @zFernand0, thanks for this new extender functionality. As regression testing with an extender that logs into it own rest api worked fine but when I tried with zosmf / apiml combo I got this error that I don't see with released ZE using same profile Screen Shot 2023-12-15 at 9 45 51 AM

This PR will fix the issue :yum:
- https://github.com/zowe/zowe-cli/pull/2014

**Comment by JillieBeanSim** [comment]
thanks @zFernand0, can re-request review once pulling in those dep updates.

**Comment by zFernand0** [comment]
Note:
- The version that contains the fix for this PR is 7.21.4
- This also contains the updates made by @KevinLoesch1
  - These updates fix #2610, 
  Thanks Kevin!

**Comment by zFernand0** [comment]
The Lint failures seem unrelated since those were working before updating from `main`
- https://github.com/zowe/vscode-extension-for-zowe/actions/runs/7201520591/job/19617824160

File: packages/zowe-explorer/CHANGELOG.md [hunk 1]
```diff
@@ -6,8 +6,10 @@ All notable changes to the "vscode-extension-for-zowe" extension will be documen
 
 ### New features and enhancements
 
+
```
Comment (traeok) [inline packages/zowe-explorer/CHANGELOG.md:9]: I think this extra blank line is causing the Lint stage to fail 😋

File: packages/zowe-explorer-api/CHANGELOG.md [hunk 2]
```diff
@@ -6,11 +6,14 @@ All notable changes to the "zowe-explorer-api" extension will be documented in t
 
 ### New features and enhancements
 
+
```
Comment (traeok) [inline packages/zowe-explorer-api/CHANGELOG.md:9]: I see there's an extra blank space here as well 😅

**Review by traeok — COMMENTED** [review]
Logic looks good! Going to test shortly. Thanks @zFernand0!
In the meantime, I think I found the source of the lint stage failure - I added comments for those lines.

File: packages/zowe-explorer/__tests__/__unit__/Profiles.extended.unit.test.ts [hunk 3]
```diff
@@ -1665,7 +1665,12 @@ describe("Profiles Unit Tests - function ssoLogout", () => {
             globalMocks.testSession,
             globalMocks.testProfile
         );
-        testNode.profile.profile.password = undefined;
+        // testNode.profile.profile.password = undefined;
+        delete testNode.profile.profile.password;
+        // Note: The above needs to be deleted instead of set to undefined.
+        //      Otherwise the `ProfilesUtils.isProfileUsingBasicAuth` reports that we are using basic auth
+        //      because "password" in testNode.profile.profile is "true" even though it's undefined
```
Comment (t1m0thyj) [inline packages/zowe-explorer/__tests__/__unit__/Profiles.extended.unit.test.ts:1672]: Is the check in `ProfilesUtils.isProfileUsingBasicAuth` behaving as expected? Curious if this:
`"user" in prof && "password" in prof;`
should be changed to:
`prof.user != null && prof.password != null;`

**Review by t1m0thyj — CHANGES_REQUESTED** [review]
Looks pretty good, thanks @zFernand0 😺

When testing this PR, I noticed there is text pre-filled in the input boxes for user/password. I think we want to retain the previous behavior of placeholder values that are "grayed out".

**Before**
image
image

**After**
image
image

File: packages/zowe-explorer/__tests__/__unit__/Profiles.extended.unit.test.ts (hunk 3 above)
Comment (zFernand0) [reply packages/zowe-explorer/__tests__/__unit__/Profiles.extended.unit.test.ts:1672]: Opened: 
- https://github.com/zowe/vscode-extension-for-zowe/issues/2664

File: packages/zowe-explorer-api/src/profiles/ProfilesCache.ts [hunk 4]
```diff
@@ -548,4 +548,25 @@ export class ProfilesCache {
             profile?.profile.tokenType === zowe.imperative.SessConstants.TOKEN_TYPE_APIML
```
[the remark this answers is not in the document]
Comment (zFernand0) [reply packages/zowe-explorer-api/src/profiles/ProfilesCache.ts:548]: Opened:
- https://github.com/zowe/vscode-extension-for-zowe/issues/2665

**Comment by zFernand0** [comment]
> LGTM! thanks zFernand0 for this update

Thank you, @JillieBeanSim for updating the package i18n 😋

**Review by traeok — APPROVED** [review]
LGTM! Thanks @zFernand0

**Comment by zFernand0** [comment]
The code duplication seems to be on new tests added.
Thanks for reviewing! 😋 
Merging!

**Pull request merged**